### DISCUSSION TITLE
JSON-LD function similar to generateMetadata

### DISCUSSION THREAD

**Discussion by davin-young** [vercel/next.js#56420] [Ideas]
### Goals

1. Should provide an easier way to generate JSON LD structured data in the head element. using generateStructuredData 
2. Should be rendered into the head without any javascript for google rich results
3. Stop the use of dangerouslySetInnerHTML
```html
 <script
        type="application/ld+json"
        dangerouslySetInnerHTML={{ __html: JSON.stringify(jsonLd) }}
      />
``` 

### Background

The current documentation for JSON-LD on the website doesnt work when pushed to production, as it needs to be wrapped in a vanilla <head> tag, which is not stated.
 
``` javascript
export default async function Page({ params }) {
  const product = await getProduct(params.id)
 
  const jsonLd = {
    '@context': 'https://schema.org',
    '@type': 'Product',
    name: product.name,
    image: product.image,
    description: product.description,
  }
 
  return (
    <section>
      {/* Add JSON-LD to your page */}
      <script
        type="application/ld+json"
        dangerouslySetInnerHTML={{ __html: JSON.stringify(jsonLd) }}
      />
      {/* ... */}
    </section>
  )
}
```

When the web page is pushed to production, [Schema Markup Validator](https://validator.schema.org/) passes and [Google Rich Results](https://search.google.com/test/rich-results) only sees
```html
<script>
    self.__next_f.push([
        1,
        '12:["$","section",null,{"children":[["$","head",null,{"children":["$","script",null,{"type":"application/ld+json","dangerouslySetInnerHTML":{".... rest of json ld
    ]);
</script>
```

However if you copy and paste the entire website html code from the browsers development tools into the code test on rich results it DOES pass as the javascript has had time to run.

### Proposal

I believe copying a similar structure to generateMetadata  will work for generateStructuredData and adding in schema-dts as a dependency for next would be cool too.

**Comment by Illyism** [comment]
Seeing how important this is for SEO, this should be a 1st class feature

**Comment by pranavks** [comment]
+1

Although I'd prefer adding the schema functionality in the existing generateMetadata function itself to make things simpler.

In our codebase, in every page we use generateMetadata and inside it we have a helper function that handles metadata that comes from a CMS. To setup schema markup, as of now we have to remember to add the script tag for JSON-LD inside the Page component in all pages. If we can get it done inside the generateMetadata function itself it will make all our pages consistent with SEO practices and we'll have to deal with less code.

**Reply by davin-young** [reply]
I like that idea too. Originally my thoughts were to keep them separate as they were different schemas, but to simplify it to one json object would be an interesting idea.

**Reply by kweheliye** [reply]
+1  adding schema functionality to generateMetadata function would help and make things easier

**Reply by WalterOkumu** [reply]
I have something similar, I am dynamically generating metadata using generateMetadata, I was hoping to simply create a function and pass variables to create schema

**Comment by hungnv1312** [comment]
yes, I fixed it use if else if own page res schema Markup json from server to client

**Reply by cap-mfm** [reply]
?

**Comment by don-esteban** [comment]
I tested the [official solution](https://nextjs.org/docs/app/building-your-application/optimizing/metadata#json-ld). Everthing seem doing well. I did a first quick check with `curl http://localhost:3000/` in local dev and production mode and I deployed the site to Vercel.

Enhanced support, similar to what generateMetadata offers, would be nice, but it is merely a nice-to-have.

**Comment by saldomik** [comment]
Can not putting the Json/LD script inside the head tag of the page cause SEO problems?

**Reply by don-esteban** [reply]
Check in Google Search Console or in another tool, if your schema.org is recognized. It does not has to be in the `<head>`.

**Comment by modernxpunk** [comment]
any workarounds?

**Reply by philwolstenholme** [reply]
https://nextjs.org/docs/app/guides/json-ld is the current recommendation
